pkg/pillar/zedcloud/send.go

```diff
@@ -614,6 +614,127 @@ func SendOnIntf(ctx *ZedCloudContext, destURL string, intf string, reqlen int64,
+		reqURL = destURL
+	} else {
+		if strings.HasPrefix(destURL, "https:") {
+			reqURL = destURL
+		} else {
+			reqURL = "https://" + destURL
+		}
+	}
+
+	if b == nil {
+		isGet = true
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: ctx.TlsConfig,
+	}
+	// Since we recreate the transport on each call there is no benefit
+	// to keeping the connections open.
+	defer transport.CloseIdleConnections()
+
+	// Try all addresses
+	localTCPAddr := net.TCPAddr{IP: ipSrc}
+	localUDPAddr := net.UDPAddr{IP: ipSrc}
+	resolverDial := func(ctx context.Context, network, address string) (net.Conn, error) {
+		log.Tracef("resolverDial %v %v", network, address)
+		// XXX can we fallback to TCP? Would get a mismatched address if we do
+		d := net.Dialer{LocalAddr: &localUDPAddr}
+		return d.Dial(network, address)
+	}
+	r := net.Resolver{Dial: resolverDial, PreferGo: true,
+		StrictErrors: false}
+	d := net.Dialer{Resolver: &r, LocalAddr: &localTCPAddr}
+	transport.Dial = d.Dial
+
+	client := &http.Client{Transport: transport}
+	if ctx.NetworkSendTimeout != 0 {
+		client.Timeout = time.Duration(ctx.NetworkSendTimeout) * time.Second
+	}
+
+	var req *http.Request
+	var err error
+
+	if b != nil {
+		req, err = http.NewRequest("POST", reqURL, b)
+	} else {
+		req, err = http.NewRequest("GET", reqURL, nil)
+	}
+	if err != nil {
+		return nil, nil, fmt.Errorf("NewRequest failed %s", err)
+	}
+
+	// Add a per-request UUID to the HTTP Header
+	// for tracability in the receiver
+	req.Header.Add("X-Request-Id", uuid.NewV4().String())
+
+	trace := &httptrace.ClientTrace{
+		GotConn: func(connInfo httptrace.GotConnInfo) {
+			log.Tracef("Got RemoteAddr: %+v, LocalAddr: %+v\n",
+				connInfo.Conn.RemoteAddr(),
+				connInfo.Conn.LocalAddr())
+		},
+		DNSDone: func(dnsInfo httptrace.DNSDoneInfo) {
+			log.Tracef("DNS Info: %+v\n", dnsInfo)
+		},
+		DNSStart: func(dnsInfo httptrace.DNSStartInfo) {
+			log.Tracef("DNS start: %+v\n", dnsInfo)
+		},
+	}
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(),
+		trace))
+	log.Tracef("SendLocal: req method %s, isget %v, url %s",
+		req.Method, isGet, reqURL)
+	callStartTime := time.Now()
+	resp, err := client.Do(req)
+	if err != nil {
+		errStr := fmt.Sprintf("client.Do (timeout %d) fail: %v", ctx.NetworkSendTimeout, err)
+		log.Errorln(errStr)
+		if ctx.FailureFunc != nil {
+			ctx.FailureFunc(log, intf, reqURL, reqlen, 0, false)
+		}
+		return nil, nil, errors.New(errStr)
+	}
+
+	contents, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		resp.Body.Close()
+		resp.Body = nil
+		if ctx.FailureFunc != nil {
+			ctx.FailureFunc(log, intf, reqURL, reqlen, 0, false)
+		}
+		return nil, nil, fmt.Errorf("ReadAll failed: %v", err)
+	}
+	resp.Body.Close()
+	resplen := int64(len(contents))
+	resp.Body = nil
+
+	totalTimeMillis := int64(time.Since(callStartTime) / time.Millisecond)
+	if ctx.SuccessFunc != nil {
+		ctx.SuccessFunc(log, intf, reqURL, reqlen, resplen, totalTimeMillis)
+	}
```

Comment:
Needs to be inside case, since otherwise you will record a failure and success for the same attempt wih a bad resp.StatusCode.